components/formats-gpl/src/loci/formats/in/BDVReader.java

```diff
@@ -0,0 +1,713 @@
+
+import ch.systemsx.cisd.base.mdarray.MDIntArray;
+import ch.systemsx.cisd.hdf5.HDF5CompoundDataMap;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.xml.sax.Attributes;
+import org.xml.sax.helpers.DefaultHandler;
+
+import loci.common.DataTools;
+import loci.common.Location;
+import loci.common.RandomAccessInputStream;
+import loci.common.services.DependencyException;
+import loci.common.services.ServiceFactory;
+import loci.common.xml.BaseHandler;
+import loci.common.xml.XMLTools;
+import loci.formats.CoreMetadata;
+import loci.formats.FormatException;
+import loci.formats.FormatReader;
+import loci.formats.FormatTools;
+import loci.formats.MetadataTools;
+import loci.formats.MissingLibraryException;
+import loci.formats.meta.MetadataStore;
+import loci.formats.services.JHDFService;
+import loci.formats.services.JHDFServiceImpl;
+
+import ome.xml.model.primitives.Color;
+import ome.xml.model.primitives.NonNegativeInteger;
+
+/**
+ * Reader for BDV files.
+ */
+public class BDVReader extends FormatReader {
+
+  // -- Constants --
+
+  public static final String HDF_MAGIC_STRING = "HDF";
+  public static final String BDV_MAGIC_STRING = "SpimData";
+  // Default colors for bounding box colors if no classification present
+  private final int[][] COLORS = {{255, 0, 0}, {0, 255, 0}, {0, 0, 255},
+                              {255, 255, 0}, {0, 255, 255}, {255, 0, 255},
+                              {255, 255, 255}, {255, 0, 128}, {0, 255, 128},
+                              {0, 128, 256}, {128, 0, 128}, {255, 128, 0},
+                              {64, 128, 0}, {0, 64, 128}, {128, 0, 64}};
+
+  // -- Fields --
+
+  private MetadataStore store;
+  private String bdvxml;
+  String xmlId;
+  String h5Id;
+  private int sizeC = 0;
+  private Integer firstTimepoint = 0;
+  private Integer lastTimepoint = 0;
+  private int seriesCount;
+  private transient JHDFService jhdf;
+  private int lastChannel = 0;
+
+  private List<H5Coordinate> H5PositionList = new ArrayList<H5Coordinate>();
+  private List<String> H5PathsToImageData = new ArrayList<String>();
+  private List<String> cellObjectNames = new ArrayList<String>();
+
+  private HDF5CompoundDataMap[] times = null;
+  private HDF5CompoundDataMap[] classes = null;
+  private HDF5CompoundDataMap[] bbox = null;
+
+  // -- Constructor --
+
+  /**
+   * Constructs a new BDV reader.
+   */
+  public BDVReader() {
+    super("BDV", "xml");
+    suffixSufficient = false;
+    domains = new String[] {FormatTools.UNKNOWN_DOMAIN};
+  }
+
+  // -- Internal FormatReader API methods --
+
+  /* @see loci.formats.FormatReader#initFile(String) */
+  @Override
+  protected void initFile(String id) throws FormatException, IOException {
+    super.initFile(id);
+    xmlId = id;
+    store = makeFilterMetadata();
+    in = new RandomAccessInputStream(xmlId);
+    in.setEncoding("ASCII");
+    DefaultHandler handler = new BDVXMLHandler();
+    try {
```

Comment:
This could leak, change recommended
```
try (RandomAccessInputStream s = new RandomAccessInputStream(id)) {
      XMLTools.parseXML(s, handler);
 } catch (IOException e) {
      throw new FormatException("Malformed XML", e);
 } 

```

Reply:
Perhaps something that could be part of XMLTools?

Reply:
From previous discussions I understood that the reader is responsible to close the stream. So I   did not look at changing the underline methods. TIFFParser for example follows the same pattern

Reply:
I was thinking that another `parseXML` method would take the `id` and then could manage the stream internally.